Type Constructors for Generic Directives

zarend opened the pull request:
## Background Knowledge Required
This requires a bit of background knowledge
 - How to use bound generics in typescript
 - How to create a Directive in angular
 - [type-check-block (tcb) generation](https://github.com/angular/angular/pull/41004)
 -- type constructors
 -- `TcbOp`s
 -- inlining, specifically of type constructors
 

## Motivation

This PR came from an edge case with generic directives – specifically directives with bound generic parameters.

Creating a generic directive with a bound parameters creates an edge case for TCB generation. Let's see the behavior in VSCode for this situation.

In this screenshot, we have a two confusing error messages.

1. `ng serve` reports that we cannot bind to `notAnInput` because that property does not exist. This error is expected, but it doesn't show up in the IDE.

2. the language service threw an exception with `This component requires inline type checking which is not available in this environment.`

<img width="1357" alt="Screen Shot 2021-03-05 at 9 31 03 AM" src="https://user-images.githubusercontent.com/7720245/110152577-c6466280-7d96-11eb-839e-0c6b08651312.png">

The type generation code cannot inline the type constructor for this component, which causes it to throw an exception before finishing type checking the template. This means that part of the template will not be typechecked and VSCode will not be able to present diagnostics in the skipped sections.

### Why bound generics are a "no-go" for compiling TCBs

Bound generics can reference private members which are impossible for the compiler to import into the tcb. They can also contain other things such as symbols which would be impossible to import.

Application Code
```
interface PrivateInterface{}

@Componet({ selector: 'bad' })
export class BadComponent <T extends PrivateInterface>{}
```

TCB bad.component.typecheck.ts
```
  var ctor1: ()<T extends PrivateInterface> => BadComponent<T> = () => (null!);
  // but wait, how do we import PrivateInterface?
```

The solutions approach is to assign `any` to generic arguments for constructors with bound parameters.

```
var ctor1: BadComponent<any> = (null!);
```

### Why inline type constructors impact performance.

When we compile a program that requires inlining type constructors, we change add generated code to the component file that the user originally wrote.

bad.component.ts
```
...
export class BadComponent <T extends PrivateInterface>{
...
// type constructor will look something like this, but not exactly like this
ngTypeConstructor: ()<T> => BadComponent<T> = () => (null!) 
}
```

Changing this file can trigger a second compile, which increases the total compilation time.

Also, we cannot change a user's file while it's been edited, which is why inlining is not available for the language service. It's only available when we compile a build target for the program.

## Solution
This is a 2-part PR.

The first commit refactors our tests, which allows use to use TDD in the next commit, which fixes the bug.

### Part 1: refactor tests

For the tests in //packages/compiler-cli/src/ngtsc/typecheck, this
commits uses a `TypeCheckFile` for the environment, rather than a
`FakeEnvironment`. Using a real environment gives us more flexibility
with testing.

### Part 2: implement fix for generic directives with bound parameters

This commit fixes the behavior when creating a type constructor for a directive when the following conditions are met.
1. The directive has bound generic parameters.
2. Inlining is not available. (This happens for language service compiles).

Previously, we would throw an error saying 'Inlining is not supported in this environment.' The compiler would stop type checking, and the developer could lose out on getting errors after the compiler gives up.

This commit adds a `useInlineTypeConstructors` to the type check config. When set to false, we use `any` type for bound generic parameters to avoid crashing. When set to true, we inline the type constructor when inlining is required.

zarend committed:
refactor(compiler-cli): don't use FakeEnvironment for tcb tests

For the tests in //packages/compiler-cli/src/ngtsc/typecheck, this
commits uses a `TypeCheckFile` for the environment, rather than a
`FakeEnvironment`. Using a real environment gives us more flexibility
with testing.

zarend committed:
fix(compiler-cli): add `useInlining` option to type check config

This commit fixes the behavior when creating a type constructor for a directive when the following
conditions are met.
1. The directive has bound generic parameters.
2. Inlining is not available. (This happens for language service compiles).

Previously, we would throw an error saying 'Inlining is not supported in this environment.' The
compiler would stop type checking, and the developer could lose out on getting errors after the
compiler gives up.

This commit adds a useInlineTypeConstructors to the type check config. When set to false, we use
`any` type for bound generic parameters to avoid crashing. When set to true, we inline the type
constructor when inlining is required.

Addresses #40963

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts:
```diff
@@ -1077,4 +1078,36 @@ describe('type check blocks', () => {
+    `;
+       const declarations: TestDeclaration[] = [{
+         code: `
+        interface PrivateInterface{}; export class Dir<T extends PrivateInterface, U extends string> {};
```
```suggestion
        interface PrivateInterface{}; 
        export class Dir<T extends PrivateInterface, U extends string> {};
```

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/test/test_utils.ts:
```diff
@@ -188,6 +188,11 @@ export const ALL_ENABLED_CONFIG: TypeCheckingConfig = {
+  useInlineTypeConstructors: true
+};
+
+export type TestBase = {
```
This type `TestBase` and the property `code` are somewhat poorly named, especially without documentation. I needed to actually find where code was used to figure out what it did. Also, why not just add the new `code`/`customCode` to the existing type rather than creating a new one and doing `&TestBase`?

alxhub replied:
Also both this and `TestPipe`/`TestDirective` should arguably be interfaces, not `type`s.

zarend replied:
done,  TestPipe/TestDirective are interfaces now, and there is no more `TestBase`

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -271,10 +266,6 @@ export class TypeCheckContextImpl implements TypeCheckContext {
         shimData.oobRecorder.requiresInlineTcb(templateId, ref.node);
       }
 
-      if (missingInlines.length > 0) {
```
Shouldn't we still be keeping track of these? Wouldn't we still want to produce an informational diagnostic to indicate that we don't have the correct types but are instead using `any` in these cases?

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/api/api.ts:
```diff
@@ -260,6 +260,16 @@ export interface TypeCheckingConfig {
+   * If false, never use inline type constructors. If a type constructor would normally require
+   * inlining, use `any` type instead to avoid inlining.
+   */
+  useInlineTypeConstructors: boolean;
```
I'm a bit confused with how this interacts with `InliningMode`. If `useInlineTypeConstructors` is `true` and `InliningMode` is `Error`, what do you expect to happen when something requires inlining? If that's not a valid configuration, then shouldn't we just have one way to configure inlining rather than two that can be at odds with each other?

atscott reviewed (changes requested):
You have a pretty good overview of the issue in the comment. Please add a similar description to the commit message. The commit messages should contain all the same information about why a change is being made - it shouldn't only be in the PR comment.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1383,8 +1429,21 @@ class Scope {
+      const host = this.tcb.env.reflector;
+      const dirRef = dir.ref as Reference<ClassDeclaration<ts.ClassDeclaration>>;
+
+      if (dir.isGeneric) {
```
The flow of this `if` structure is very unintuitive, and I think it could be made a lot cleaner and more readable by putting the cases in the order of commonality:

```
if (!dir.isGeneric) {
  // TcbDirectiveTypeOp
} else if (useInlineTypeCtors || !requiresInLineTypeCtor) {
  // TcbDirectiveCtorOp
} else {
  // fallback op (name subject to bikeshedding)
}
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_file.ts:
```diff
@@ -49,12 +49,12 @@ export class TypeCheckFile extends Environment {
   }
 
-  render(): string {
+  render(options?: {removeComments?: boolean}): string {
```
Yeah, I think required here would be better.

zarend replied:
fixed: make this a required param

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -690,6 +690,7 @@ export class NgCompiler {
         useContextGenericType: strictTemplates,
         strictLiteralTypes: true,
         enableTemplateTypeChecker: this.enableTemplateTypeChecker,
+        useInlineTypeConstructors: false,
```
This should definitely not be `false`. Inline type constructors are required for how the compiler normally operates.

This should probably be set based on whether the current `TypeCheckingProgramStrategy` supports inline operations (it exposes a `supportsInlineOperations` property to report this).

zarend replied:
🤦 , that is correct.

For strict mode, I made it
 `useInlineTypeConstructors: this.typeCheckingProgramStrategy.supportsInlineOperations,`
That's because

For non-strict mode, I set it to false. That's because we want to be as lenient as possible in non-strict mode. When `useInlineTypeConstructors` is `false`, we use any for the generic params of bound generics instead of trying to infer the types. It's better to avoid inlining unless we have to because it's problematic ( perf issues, complications with editing user's files).

zarend commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -390,6 +391,51 @@ class TcbDirectiveTypeOp extends TcbOp {
+    }
+
+    const typeArguments =
+        dirRef.node.typeParameters!.map(() => ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword));
```
Good suggestions, I made this an explicit check-and-throw. I like doing a check-and-throw instead of a non-null assertion, so that it doesn't look like a mistake to the reader of the code. In this example, null really is an invalid state as this Op is only intended for generic directives.

zarend commented:
@alxhub @atscott , I've addressed the PR comments, and this PR is ready for your eyes again. 

(I still need to update the commit description, will take care of that this morning).

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -689,6 +689,7 @@ export class NgCompiler {
         useContextGenericType: strictTemplates,
         strictLiteralTypes: true,
         enableTemplateTypeChecker: this.enableTemplateTypeChecker,
+        useInlineTypeConstructors: this.typeCheckingProgramStrategy.supportsInlineOperations,
```
This should probably also take into consideration the `InliningMode` - if inlining is disabled, we should honor that even if the type-checking strategy supports it.

zarend replied:
Hmm... I think we're mixing up `InliningMode` and `supportsInlineOperations`.

`supportsInlineOperations` it's a parameter of the `TypecheckContextImpl`. It is `false` for the language service, since we cannot edit working users files. It is `true` while compiling for real.

`InliningMode` is an enum that says how the type-checking context handles inlining. 

```
export enum InliningMode {
  /**
   * Use inlining operations when required.
   */
  InlineOps, // <- use this when `supportsInlineOperations` is true.

  /**
   * Produce diagnostics if an operation would require inlining.
   */
  Error, // use this when `supportsInlineOperations` is false.
}
```

The InliningMode is only used only by the type-checking context, and it is set based on the `supportsInlineOperations`. This is handled in `TemplateTypechecker::newContext`.

https://github.com/zarend/angular/blob/master/packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts#L397

Since the source of truth for InliningMode is `supportsInlineOperations`. We only need to check `supportsInlineOperations` here, and we don't need to know about `InliningMode`.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -390,6 +391,54 @@ class TcbDirectiveTypeOp extends TcbOp {
+      throw new Error(
+          `Expected TypeReferenceNode when referencing the type for ${this.dir.ref.debugName}`);
+    }
+    if (!dirRef.node.typeParameters) {
```
Explicitly test for `undefined` and don't rely on truthiness.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1383,8 +1432,20 @@ class Scope {
+      const host = this.tcb.env.reflector;
+      const dirRef = dir.ref as Reference<ClassDeclaration<ts.ClassDeclaration>>;
+
+      if (!dir.isGeneric) {
```
This construct could benefit from some inline comments, describing the logic behind each case.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1383,8 +1432,20 @@ class Scope {
+
+      if (!dir.isGeneric) {
+        directiveOp = new TcbDirectiveTypeOp(this.tcb, this, node, dir);
+      } else if (
```
I also suggest inverting this condition and switching it with the one below, so the logic proceeds according to most common -> least common.

That is, the cases are:
* Literal type is okay (most likely)
* Requires a type constructor, and we can generate one (less likely)
* Would require a type constructor, but we have to fall back on a literal type with anys due to inlining restrictions (least likely)
 
This helps to make the logic easier to read and reason about.

zarend replied:
done, I added inline comments, and reordered the branches. It turned out nicely, it looks less complicated after doing this 💯

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -367,8 +368,8 @@ class TcbTextInterpolationOp extends TcbOp {
   constructor(
-      private tcb: Context, private scope: Scope, private node: TmplAstTemplate|TmplAstElement,
-      private dir: TypeCheckableDirectiveMeta) {
+      protected tcb: Context, protected scope: Scope,
```
Ah, I do like how the inheritance lets us share a bunch of stuff. Civen that our decision was that the new Op and this existing one serve separate purposes, it feels a bit odd that the new Op would extend the existing one. Would it be better to have an abstract base class?

```
abstract class X extends TcbOp {
...
 abstract execute(): ts.Identifier;
...
}

class TcbDirectiveTypeOp extends X {
  execute(): ts.Identifier {
    const type = this.tcb.env.referenceType(this.dir.ref);
    return this.createVariableDeclaration(type);
  }
}

class TcbGenericDirectiveTypeWithAnyParamsOp extends X {
  execute() {...}
}
```
@alxhub

zarend replied:
I don't think an abstract base class is necessary, but I don't feel strongly about it :).

Extending `TcbDirectiveTypeOp` seems natural to me, because the new op does the exact same thing, except it also "fills in" the generic parameters.

In other works. `TcbDirectiveTypeOp` is the basic version which makes a tcb that looks like this.
`var dir1: i0.Dir = (null!); //psuedo code, actual tcb may vary`

and `TcbGenericDirectiveTypeWithAnyParamsOp`, builds on the previous op by filling in the `<>`
`var dir1: i0.Dir<any,any> = (null!); //psuedo code, actual tcb may vary`

Potentially, we could extend `TcbDirectiveTypeOp` again to make another op which tries to use actual types instead of anys. (not a real use case right now, but the point is that `TcbDirectiveTypeOp` is extendable.
`var dir1: i0.Dir<string,number> = (null!); //psuedo code, actual tcb may vary`

zarend replied:
Would it help if I wrote a comment for `TcbDirectiveTypeOp` that says it's extensible? That makes it's not ambiguous which op's are intended to be extended or not.

alxhub replied:
I don't think either option is right, personally. Extension implies substitutability, which is not true in this case because `TcbGenericDirectiveTypeWithAnyParamsOp` is semantically different than `TcbDirectiveTypeOp` and cannot be used in its place. Instead, we should put the logic either in an abstract base class or in a shared utility function.

zarend replied:
okay, I changed it to use a helper function.

zarend replied:
after further discussion over github and offline, implemented this using an abstract class – see comments below

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -219,10 +219,6 @@ export class TypeCheckContextImpl implements TypeCheckContext {
 
-    // Accumulate a list of any directives which could not have type constructors generated due to
-    // unsupported inlining operations.
-    let missingInlines: ClassDeclaration[] = [];
```
We should, however, throw an exception here if the `InliningMode` is `Error`, but `config.useInlineTypeConstructors` is `true`.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -356,6 +357,20 @@ class TcbTextInterpolationOp extends TcbOp {
+ * Creates a variable declaration of the argument type and adds it to the tcb. Returns the id of the
+ * newly created variable.
+ */
+function createVariableDeclarationOfType(
```
This name is fairly generic for its purpose, which specifically handles a variable for a directive type on a `TmplAstTemplate|TmplAstElement` node.

Additionally, I think a lot more of the logic could be shared.

Think about it this way: both operations do the same thing from a structural standpoint. The only difference is whether type parameters are allowed or not.

I think we should use an abstract base class `TcbOp`, which performs the logic:

* import the directive raw type
* if the directive has generic parameters, build a type that has `any`s
* tag it appropriately
* return the id

Then, both `TcbDirectiveTypeOp` and `TcbGenericDirectiveTypeWithAnyParamsOp` can extend from the base version, and specialize the `execute()` operation:

* `TcbDirectiveTypeOp` will assert that the directive class is _not_ generic, before delegating to `super.execute()`.
* `TcbGenericDirectiveTypeWithAnyParamsOp` will assert that the directive class _is_ generic.

zarend replied:
done ✔️

zarend commented on packages/language-service/ivy/test/quick_info_spec.ts:
```diff
@@ -479,6 +479,49 @@ describe('quick info', () => {
+      env = LanguageServiceTestEnv.setup();
+    });
+
+    it('should get quick info for the generic input of a directive that normally requires inlining',
```
Note to reviewers: this tests the original bug that we're trying to fix :). Language service crashes when encountering directives that would normally be inlined. This makes some IDE features not work such as quick info, and causes some of the developer's code not to be type checked.

zarend commented:
@atscott @alxhub the PR comments have been fixed, and this is ready for your eyes again.

Since you last reviewed, I've used an abstract class to implement TcbOb's, updated the comments for `useInlineTypeConstructors`, made it an error to set `useInlineTypeConstructors` to `true` when the inlining mode is `InliningMode.Error`, and made a few other misc fixes.

atscott reviewed (approved):
👌  I like how clean the abstract base class ends up being. Good work!

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -179,7 +179,12 @@ export class TypeCheckContextImpl implements TypeCheckContext {
       private refEmitter: ReferenceEmitter, private reflector: ReflectionHost,
-      private host: TypeCheckingHost, private inlining: InliningMode) {}
+      private host: TypeCheckingHost, private inlining: InliningMode) {
+    if ((inlining === InliningMode.Error) && config.useInlineTypeConstructors) {
```
Not: no need for the extra paren wrapping here.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -262,7 +265,7 @@ export class TypeCheckContextImpl implements TypeCheckContext {
     // If inlining is not supported, but is required for either the TCB or one of its directive
     // dependencies, then exit here with an error.
-    if (this.inlining === InliningMode.Error && (tcbRequiresInline || missingInlines.length > 0)) {
+    if (this.inlining === InliningMode.Error && tcbRequiresInline) {
```
Having `tcbRequiresInline` here means that the inner `if` on line 273 is redundant.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1,4 +1,4 @@
-/**
+/*
```
?

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -357,18 +358,13 @@ class TcbTextInterpolationOp extends TcbOp {
+ * parameters are set to `any` type.
  */
-class TcbDirectiveTypeOp extends TcbOp {
+abstract class TcbDirectiveTypeBaseOp extends TcbOp {
```
```suggestion
abstract class TcbDirectiveTypeOpBase extends TcbOp {
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -380,16 +376,82 @@ class TcbDirectiveTypeOp extends TcbOp {
+   * Creates a variable declaration for this directive of the argument type. Adds the variable
+   * declaration to the tcb and returns the id of the newly created variable.
+   */
+  private createVariableDeclarationOfType(type: ts.TypeNode): ts.Identifier {
```
There's only one caller here - shouldn't we just inline this?

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -380,16 +376,82 @@ class TcbDirectiveTypeOp extends TcbOp {
+   */
+  execute(): ts.Identifier {
+    const dirRef = this.dir.ref as Reference<ClassDeclaration<ts.ClassDeclaration>>;
+    if (this.dir.isGeneric === true) {
```
```suggestion
    if (this.dir.isGeneric) {
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -380,16 +376,82 @@ class TcbDirectiveTypeOp extends TcbOp {
+class TcbGenericDirectiveTypeWithAnyParamsOp extends TcbDirectiveTypeBaseOp {
+  execute(): ts.Identifier {
+    const dirRef = this.dir.ref as Reference<ClassDeclaration<ts.ClassDeclaration>>;
+    if (dirRef.node.typeParameters === undefined) {
```
Should we check `!this.dir.isGeneric` too?

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts:
```diff
@@ -1063,9 +1073,42 @@ describe('type check blocks', () => {
+         isGeneric: true
+       }];
+
+       const config: Partial<TypeCheckingConfig> = {useInlineTypeConstructors: false};
```
Probably don't need to have this declared as a separate constant.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/test/test_utils.ts:
```diff
@@ -354,7 +393,13 @@ export function setup(targets: TypeCheckingTarget[], overrides: {
+
+  const defaultConfig = ALL_ENABLED_CONFIG;
+  if (overrides.inlining !== undefined) {
+    defaultConfig.useInlineTypeConstructors = overrides.inlining;
```
This is overwriting a constant. We should set this on `fullConfig` below instead.

mhevery commented:
@zarend This had a merge conflict on patch branch. Please create a new PR and label accordingly

zarend commented:
> @zarend This had a merge conflict on patch branch. Please create a new PR and label accordingly

Thanks, here's a PR to merge it into patch branch.
#41268
